[dynamo] Fix frozen dataclass reconstruction

desertfire opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #169614

Summary: Fixes https://github.com/pytorch/pytorch/issues/169489.

When using `torch.compile` with `torch.utils.checkpoint.checkpoint` on a function that returns a pytree-registered frozen dataclass, compilation failed because `FrozenDataClassVariable` lacked a proper `reconstruct` method for sourceless variables. This PR implements the `reconstruct` method .

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @chauhang @amjames @Lucaskabela @jataylo @mlazos

desertfire committed:
[dynamo] Fix frozen dataclass reconstruction

Summary: Fixes https://github.com/pytorch/pytorch/issues/169489.

When using `torch.compile` with `torch.utils.checkpoint.checkpoint` on a function that returns a pytree-registered frozen dataclass, compilation failed because `FrozenDataClassVariable` lacked a proper `reconstruct` method for sourceless variables. This PR implements the `reconstruct` method .

[ghstack-poisoned]

Lucaskabela commented on torch/_dynamo/variables/user_defined.py:
```diff
@@ -1855,8 +1857,58 @@ def reconstruct(self, codegen: "PyCodegen") -> None:
+                continue
+            field_vt = self.fields.get(field.name)
+            if field_vt is None:
+                # Field not set - this shouldn't happen for a properly constructed
```
Since this shouldn't happen, should we think about throwing an exception here?

Lucaskabela reviewed (approved):
Minor comment on what we should do in case we get unexpected dataclass, but generally lgtm

desertfire committed:
Update on "[dynamo] Fix frozen dataclass reconstruction"


Summary: Fixes https://github.com/pytorch/pytorch/issues/169489.

When using `torch.compile` with `torch.utils.checkpoint.checkpoint` on a function that returns a pytree-registered frozen dataclass, compilation failed because `FrozenDataClassVariable` lacked a proper `reconstruct` method for sourceless variables. This PR implements the `reconstruct` method .

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo mlazos

[ghstack-poisoned]

izaitsevfb commented:
@pytorchbot revert -c nosignal -m "breaks lint"

[GH job link](https://github.com/pytorch/pytorch/actions/runs/20050465945/job/57505083981) [HUD commit link](https://hud.pytorch.org/pytorch/pytorch/commit/0880c80f603f0615e4e8a640016264ffc295eb5d)

desertfire committed:
Update on "[dynamo] Fix frozen dataclass reconstruction"


Summary: Fixes https://github.com/pytorch/pytorch/issues/169489.

When using `torch.compile` with `torch.utils.checkpoint.checkpoint` on a function that returns a pytree-registered frozen dataclass, compilation failed because `FrozenDataClassVariable` lacked a proper `reconstruct` method for sourceless variables. This PR implements the `reconstruct` method .

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo mlazos

[ghstack-poisoned]

desertfire commented:
@pytorchbot merge
